### PR TITLE
Automatically get transaction fees

### LINKED ISSUE #165: optimum scout_transaction_fee
I don't understand the "scout_transaction_fee", what value should be set?

Mine is set at 0.001, and I am noticing loss in overall coins amount..

I don't have BNB in my account so the fee is 0.1% 

Which value should be set in scout_transaction_fee?

### PR DESCRIPTION
**Pull request by j-waters**
Closes #165

### CODE REVIEW

**Comment by rel-s** [comment]
Looks great, testing now

**Review by rel-s — APPROVED** [review]
Code looks great, I wonder if we should implement another timer to update the fees (right now they are only updated on AutoTrade init). Do they actually ever change? Probably no need to implement in this PR, what do you think @edeng23 ?

**Comment by j-waters** [comment]
@rel-s good point, the only reason they would change is if you were trading enough to [move into another fee bracket](https://www.binance.com/en/fee/schedule) - probably unlikely for our users.

An alternative, which I experimented with in #137, would be to use a caching library like [cachetools](https://pypi.org/project/cachetools/). Calls to `get_alt_tick` and `get_trade_fees`, which are unlikely to ever change but might one day, could simply be cached with a TTL of a day or so.

**Comment by j-waters** [comment]
Pretty simple to implement. I can revert if needs be

**Comment by rel-s** [comment]
Seems like a good idea to me. Could you pin the new dependency though? I think it's a good idea to start doing that, and I plan on pinning all of our current ones...

**Comment by j-waters** [comment]
@rel-s done, yeah it's a good idea

**Review by rel-s — APPROVED** [review]
Approved, testing again...

**Comment by BelphegorPrime** [comment]
> @rel-s good point, the only reason they would change is if you were trading enough to [move into another fee bracket](https://www.binance.com/en/fee/schedule) - probably unlikely for our users.


it is also possible when you pay fees with BNB and run out of it that fee percentages are changing.

**Comment by rel-s** [comment]
> it is also possible when you pay fees with BNB and run out of it that fee percentages are changing.

I couldn't find any info in the official API the merely having BNB (which is the requirement for paying a lower fee, right?) will change the result of the API call.
https://binance-docs.github.io/apidocs/spot/en/#asset-detail-user_data
Why do you think it will @BelphegorPrime?

**Comment by BelphegorPrime** [comment]
you can see here at the bottom on point [4]( https://www.binance.com/en/support/faq/115000583311-Using-BNB-to-Pay-for-Fees) that when you don´t have enough BNB to pay for trading fees, it will do as normal in the fee schedule.
question is, does the api return the a fee as in the fee schedule or with the BNB bonus? I just assume in the following with the BNB bonus, because I cpulnd´t find something in the api definition.
If it is with the BNB bonus it can change between single trades.
For example you have enough BNB to pay the fees for the sell but than you don´t have enough to buy the target coin. which can be result in less pr even negative profit.

EDIT:
there is an api to check if BNB "burn" is activated https://binance-docs.github.io/apidocs/spot/en/#get-bnb-burn-status-user_data

**Comment by j-waters** [comment]
Interesting, yeah we can use `get_bnb_burn_spot_margin`, and then multiply the fees by 0.75? Because I'm using BNB for fees, I have enough, and it's saying that my fees are `0.001`.

It doesn't matter too much if we overestimate the fees, since it will just mean the bot is more conservative.

Edit: Done

**Comment by BelphegorPrime** [comment]
> For example you have enough BNB to pay the fees for the sell but than you don´t have enough to buy the target coin. which can be result in less or even negative profit.

But this problem remains.

When BNB burn is enabled but you don`t have enough BNB to pay the fees, you are not getting the bonus for upcoming trades until you have BNB again.

Check should be something like
if bnbBurnActive == true && doesHaveEnoughBnb()
then transaction_fee *= 0.75

**Comment by BelphegorPrime** [comment]
When BNB burn is active does self.BinanceClient.get_trade_fee() return the fees already with the BNB bonus of 25% off or is it the default fee schedule traiding fee?

**Comment by j-waters** [comment]
Shouldn't be too difficult to implement, I'll look into it tomorrow. We might be over complicating things when it wouldn't make much difference to just leave the raw things but as long as we're not adding too many extra API calls we may as well do this.

**Comment by rel-s** [comment]
Agreed - solving the BNB fee issue automatically is probably the way to go here, especially since we removed the option to manually set the transaction fee.

**Comment by rel-s** [comment]
Agreed - solving the BNB fee issue automatically is probably the way to go here, especially since we removed the option to manually set the transaction fee.

**Comment by j-waters** [comment]
Okay I've added logic to check if you have enough BNB. The slight limitation is that the BNB has to be in your spot account - I keep mine in an earn account and there's no way of getting that unfortunately. Not the end of the world though since it will just mean that the bot is ever so slightly more conservative.

File: binance_trade_bot/auto_trader.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def initialize_trade_thresholds(self):
             for pair in session.query(Pair).filter(Pair.ratio.is_(None)).all():
                 if not pair.from_coin.enabled or not pair.to_coin.enabled:
                     continue
-                self.logger.info(f"Initializing {pair.from_coin} vs {pair.to_coin}", False)
```
Comment (rel-s) [inline binance_trade_bot/auto_trader.py]: I think this isn't rebased properly, why would this change?

File: binance_trade_bot/config.py [hunk 2]
```diff
@@ -24,8 +23,7 @@ def __init__(self):
         if not os.path.exists(CFG_FL_NAME):
             print("No configuration file (user.cfg) found! See README. Assuming default config...")
             config[USER_CFG_SECTION] = {}
-        else:
```
Comment (rel-s) [inline binance_trade_bot/config.py]: This as well, we added this in a recent PR, this PR shouldn't break it.

**Comment by rel-s** [comment]
@j-waters I skimmed through the code and it seems like it wasn't rebased properly. Could you fix this so we can re-test?

**Comment by j-waters** [comment]
Ah yep sorry, all fixed

File: binance_trade_bot/auto_trader.py [hunk 3]
```diff
@@ -152,10 +152,16 @@ def scout(self):
             # Obtain (current coin)/(optional coin)
             coin_opt_coin_ratio = current_coin_price / optional_coin_price
 
+            transaction_fee = self.manager.get_fee(pair.from_coin, self.config.BRIDGE, True) + self.manager.get_fee(
+                pair.to_coin, self.config.BRIDGE, False
+            )
+
+            if self.manager.get_using_bnb_for_fees():
```
Comment (rel-s) [inline binance_trade_bot/auto_trader.py]: I think this code is already inside `get_fee()`, since it will return `base_fee * 0.75` if we are using BNB; why do we need to take this into account again?

File: binance_trade_bot/binance_api_manager.py [hunk 4]
```diff
@@ -97,6 +125,17 @@ def wait_for_order(self, origin_symbol, target_symbol, order_id):
     def buy_alt(self, origin_coin: Coin, target_coin: Coin, all_tickers):
         return self.retry(self._buy_alt, origin_coin, target_coin, all_tickers)
 
+    def _buy_quantity(
```
Comment (rel-s) [inline binance_trade_bot/binance_api_manager.py:128]: Not entirely sure why we need this new function.. How is this related to automatic transaction fee fetching?

File: binance_trade_bot/binance_api_manager.py [hunk 5]
```diff
@@ -143,6 +180,12 @@ def _buy_alt(self, origin_coin: Coin, target_coin: Coin, all_tickers):
     def sell_alt(self, origin_coin: Coin, target_coin: Coin):
         return self.retry(self._sell_alt, origin_coin, target_coin)
 
+    def _sell_quantity(self, origin_symbol: str, target_symbol: str, origin_balance: float = None):
```
Comment (rel-s) [inline binance_trade_bot/binance_api_manager.py:183]: Same comment here as in `_buy_quantity`

**Review by rel-s — COMMENTED** [review]
I have a few questions about this PR, see my review. Anyways, I'm already starting to test this in my environment :)

File: binance_trade_bot/auto_trader.py (hunk 3 above)
Comment (j-waters) [reply binance_trade_bot/auto_trader.py]: Ah yep I thought I had removed that but apparently not thanks

File: binance_trade_bot/binance_api_manager.py (hunk 4 above)
Comment (j-waters) [reply binance_trade_bot/binance_api_manager.py:128]: Okay, so to determine if we have enough BNB for it to be used, we need to calculate how much the fee will be. That's done [here](https://github.com/edeng23/binance-trade-bot/pull/198/files#diff-65cde28df70e6f927481325a29395037472fa5a4b7700e7e2fdc3a411891c4c9R39). The BNB amount will be the amount we're trading, multiplied by the base fee, multiplied by 0.75 (the discount), multiplied by the conversion rate between the coin we're trading and BNB. We can then compare that BNB amount with our balance.

Getting the amount that we're trading is the reason I made new functions, so that we're not duplicating any logic, and so that changes to the formula don't have to be made in multiple places. Also, I've done a similar thing in #196.

File: binance_trade_bot/binance_api_manager.py (hunk 5 above)
Comment (j-waters) [reply binance_trade_bot/binance_api_manager.py:183]: Same reason as above

**Review by rel-s — APPROVED** [review]
Looks great, testing...

**Comment by rel-s** [comment]
Seems to work perfectly on my setup! One last thing for this PR - we should probably write something in the `README` to let users know that if they have BNB, the reduced fee will be used automatically. Perhaps @edeng23 has an idea?

**Review by edeng23 — APPROVED** [review]
